### PR TITLE
Inner Transaction Search

### PR DESCRIPTION
**Pull request by winder**
## Summary

Update queries to support returning the root transaction bytes when matching an inner transaction.

## Test Plan

New unit tests.

### CODE REVIEW

File: idb/idb.go [hunk 1]
```diff
@@ -28,6 +31,9 @@ type TxnRow struct {
 	// TxnBytes is the raw signed transaction with apply data object.
 	TxnBytes []byte
 
+	// RootTxnBytes the root transaction raw signed transaction with apply data object, only inner transactions have this.
+	RootTxnBytes []byte
```
Comment (winder) [inline idb/idb.go:35]: Coalesce could be used to return RootTxnBytes / TxnBytes in the same parameter, but that could make it easy to mix up the other TxnRow

File: api/handlers.go [hunk 2]
```diff
@@ -495,16 +495,42 @@ func (si *ServerImplementation) SearchForTransactions(ctx echo.Context, params g
...
 
+	// Used for filtering duplicates after getting results.
+	rootTxnDedupeMap := make(map[string]struct{})
+	if filter.NextToken != "" {
+		_, _, txid, err := idb.DecodeTxnRowNext(filter.NextToken)
+		if err != nil {
+			badRequest(ctx, "bad next token")
+		}
+		rootTxnDedupeMap[txid] = struct{}{}
+	}
```
Comment (winder) [inline api/handlers.go:506]: Bootstrapping the dedupe map with NextToken solves the duplicate-root-txn-while-paging problem.

File: idb/idb.go [hunk 3]
```diff
@@ -39,29 +45,51 @@ type TxnRow struct {
...
 	return
 }
 
 // TxnExtra is some additional metadata needed for a transaction.
 type TxnExtra struct {
 	AssetCloseAmount uint64 `codec:"aca,omitempty"`
-	RootTxid         string `codec:"root-txid,omitempty"`
+	// RootIntra is a string to allow distinguishing between 0 and empty.
+	RootIntra string `codec:"root-intra,omitempty"`
+	RootTxid  string `codec:"root-txid,omitempty"`
```
[the remark this answers is not in the document]
Comment (winder) [reply idb/idb.go:92]: No. Yes.

File: idb/postgres/postgres.go [hunk 4]
```diff
@@ -722,11 +733,12 @@ func (db *IndexerDb) txnsWithNext(ctx context.Context, tx pgx.Tx, tf idb.Transac
...
 		// (round,intra) descending into the past
 		tf.OffsetLT = origOLT
-		if nextround == 0 {
+		tf.MaxRound = nextround - 1
+
+		if tf.MaxRound == 0 {
 			// NO second query
 			return
 		}
-		tf.MaxRound = nextround - 1
```
[the remark this answers is not in the document]
Comment (winder) [reply idb/postgres/postgres.go:729]: The problem I saw was that when `nextround` is  1, we set `tf.MaxRound = 0`. For this parameter the zero value indicates the value isn't set, so we get a bunch of unexpected results.

I think the old code was written under the assumption that transactions on round 0 was a valid thing.

Looking at this again, I think I can change the early exit to `if nextround <= 1 { return }` to improve this check slightly.

File: idb/idb.go [hunk 5]
```diff
@@ -39,29 +45,56 @@ type TxnRow struct {
...
+	}
+
+	return base64.URLEncoding.EncodeToString(b[:]), nil
 }
 
 // DecodeTxnRowNext unpacks opaque string returned from TxnRow.Next()
-func DecodeTxnRowNext(s string) (round uint64, intra uint32, err error) {
-	var b []byte
-	b, err = base64.URLEncoding.DecodeString(s)
+func DecodeTxnRowNext(s string) (round uint64, intra uint32, rootTxid string, err error) {
```
Comment (brianolson) [inline idb/idb.go:69]: If we wanted to be clever we could decode the previous version of next-token here based on the decoded length; but I don't care enough to press it and approximately zero API paging sequences will be disrupted.

Comment (winder) [reply idb/idb.go:69]: Ok, it never occurred to me to consider someone saving the next-token string between releases. I think I'm comfortable letting that generate an error.

File: api/handlers.go [hunk 6]
```diff
@@ -495,16 +495,42 @@ func (si *ServerImplementation) SearchForTransactions(ctx echo.Context, params g
...
+
 	// Fetch the transactions
 	txns, next, round, err := si.fetchTransactions(ctx.Request().Context(), filter)
```
[the remark this answers is not in the document]
Comment (winder) [reply api/handlers.go]: done.

**Pull request merged**